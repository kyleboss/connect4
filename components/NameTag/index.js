export * from './NameTag';
export { default } from './NameTag';
