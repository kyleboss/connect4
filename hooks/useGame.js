import { useContext } from "react";

import GameContext from "@/state/GameContext";

const useGame = () => useContext(GameContext);

export default useGame;
