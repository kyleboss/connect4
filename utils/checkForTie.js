const checkForTie = (board) =>
  board.every((row) => row.every((cell) => cell !== null));

module.exports = checkForTie;
