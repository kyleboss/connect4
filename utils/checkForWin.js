const checkHorizontal = (player, board) => {
  for (let row = 0; row < board.length; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row][col + 1] === player &&
        board[row][col + 2] === player &&
        board[row][col + 3] === player
      ) {
        return true;
      }
    }
  }
  return false;
};

const checkVertical = (player, board) => {
  for (let col = 0; col < 7; col++) {
    for (let row = 0; row < board.length - 3; row++) {
      if (
        board[row][col] === player &&
        board[row + 1][col] === player &&
        board[row + 2][col] === player &&
        board[row + 3][col] === player
      ) {
        return true;
      }
    }
  }
  return false;
};

const checkDiagonal = (player, board) => {
  // Check right diagonal
  for (let row = 0; row < board.length - 3; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row + 1][col + 1] === player &&
        board[row + 2][col + 2] === player &&
        board[row + 3][col + 3] === player
      ) {
        return true;
      }
    }
  }

  // Check left diagonal
  for (let row = 3; row < board.length; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row - 1][col + 1] === player &&
        board[row - 2][col + 2] === player &&
        board[row - 3][col + 3] === player
      ) {
        return true;
      }
    }
  }

  return false;
};

const checkForWin = (player, board) =>
  checkHorizontal(player, board) ||
  checkVertical(player, board) ||
  checkDiagonal(player, board);

// Must be CommonJS to accomodate for express server.
module.exports = checkForWin;
