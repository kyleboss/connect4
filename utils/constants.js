// Game State
const GAME_STATE_WAITING = "GAME_STATE_WAITING";
const GAME_STATE_ACTIVE = "GAME_STATE_ACTIVE";
const GAME_STATE_GAME_OVER = "GAME_STATE_GAME_OVER";

// User roles
const USER_ROLE_GUEST = "USER_ROLE_GUEST";
const USER_ROLE_RED = "USER_ROLE_RED";
const USER_ROLE_YELLOW = "USER_ROLE_YELLOW";
const USER_ROLE_NONE = "USER_ROLE_NONE";
const USER_ROLES = {
  [USER_ROLE_GUEST]: "Guest",
  [USER_ROLE_RED]: "Red",
  [USER_ROLE_YELLOW]: "Yellow",
  [USER_ROLE_NONE]: "No one",
};

// URLs
const CLIENT_PORT = 3000;
const SERVER_PORT = 3001;
const CLIENT_URL = `http://localhost:${CLIENT_PORT}`;
const SERVER_URL = `http://localhost:${SERVER_PORT}`;

// Socket Events
const SOCKET_EVENT_JOIN_GAME = "SOCKET_EVENT_JOIN_GAME";
const SOCKET_EVENT_RESET_GAME = "SOCKET_EVENT_RESET_GAME";
const SOCKET_EVENT_MAKE_MOVE = "SOCKET_EVENT_MAKE_MOVE";
const SOCKET_EVENT_ASSIGN_ROLE = "SOCKET_EVENT_ASSIGN_ROLE";
const SOCKET_EVENT_UPDATE_STATE = "SOCKET_EVENT_UPDATE_STATE";
const SOCKET_EVENT_CONNECTION = "connection"; // Official socket event.
const SOCKET_EVENT_DISCONNECT = "disconnect"; // Official socket event.

// Must be CommonJS to accomodate for express server.
module.exports = {
  GAME_STATE_WAITING,
  GAME_STATE_ACTIVE,
  GAME_STATE_GAME_OVER,
  USER_ROLE_GUEST,
  USER_ROLE_RED,
  USER_ROLE_YELLOW,
  USER_ROLE_NONE,
  USER_ROLES,
  CLIENT_PORT,
  SERVER_PORT,
  CLIENT_URL,
  SERVER_URL,
  SOCKET_EVENT_JOIN_GAME,
  SOCKET_EVENT_RESET_GAME,
  SOCKET_EVENT_MAKE_MOVE,
  SOCKET_EVENT_ASSIGN_ROLE,
  SOCKET_EVENT_UPDATE_STATE,
  SOCKET_EVENT_CONNECTION,
  SOCKET_EVENT_DISCONNECT,
};
