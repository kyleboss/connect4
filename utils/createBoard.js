const createBoard = () => Array(6).fill(Array(7).fill(null));

module.exports = createBoard;
