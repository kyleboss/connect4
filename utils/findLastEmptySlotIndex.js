const findLastEmptySlotIndex = (board, columnIndex) => {
  for (let i = board.length - 1; i >= 0; i--) {
    if (!board[i][columnIndex]) {
      return i;
    }
  }
  return -1; // Indicates the column is full
};

module.exports = findLastEmptySlotIndex;
