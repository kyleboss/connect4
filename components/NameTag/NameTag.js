import useGame from "@/hooks/useGame";
import { USER_ROLES } from "@/utils/constants";
import Header from "@/components/Header";

const NameTag = () => {
  const { userRole } = useGame();

  return <Header>{USER_ROLES[userRole]}</Header>;
};

export default NameTag;
