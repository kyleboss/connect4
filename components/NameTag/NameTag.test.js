import React from "react";
import { render } from "@testing-library/react";

import NameTag from "./NameTag";
import useGame from "../../hooks/useGame";
import { USER_ROLE_RED, USER_ROLE_YELLOW } from "../../utils/constants";

jest.mock("../../hooks/useGame", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("NameTag", () => {
  it("renders the user role correctly", () => {
    useGame.mockImplementation(() => ({
      userRole: USER_ROLE_RED,
    }));

    const { getByText } = render(<NameTag />);
    expect(getByText("Red")).toBeInTheDocument();
  });

  it("matches the snapshot for a specific user role", () => {
    useGame.mockImplementation(() => ({
      userRole: USER_ROLE_YELLOW,
    }));

    const { asFragment } = render(<NameTag />);
    expect(asFragment()).toMatchSnapshot();
  });
});
