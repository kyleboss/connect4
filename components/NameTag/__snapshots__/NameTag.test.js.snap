// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameTag matches the snapshot for a specific user role 1`] = `
<DocumentFragment>
  <h1
    class="sc-f9cd8f8e-0 dcLxvF"
  >
    Yellow
  </h1>
</DocumentFragment>
`;
